saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/OpenSamlAuthenticationProvider.java

```diff
@@ -494,6 +494,52 @@ private NameID decrypt(Saml2AuthenticationToken token, EncryptedID assertion)
 		throw last;
 	}
 
+	private Map<String, Object> getAssertionAttributes(Assertion assertion) {
+		Map<String, Object> attributeMap = new LinkedHashMap<>();
+		for (AttributeStatement attributeStatement : assertion.getAttributeStatements()) {
+			for (Attribute attribute : attributeStatement.getAttributes()) {
+
+				List<Object> attributeValues = new ArrayList<>();
+				for (XMLObject xmlObject : attribute.getAttributeValues()) {
+					Object attributeValue = getXmlObjectValue(xmlObject);
+					if (attributeValue != null) {
+						attributeValues.add(attributeValue);
+					}
+				}
+				attributeMap.put(attribute.getName(), attributeValues);
+
+			}
+		}
+		return attributeMap;
+	}
+
+	private Object getXmlObjectValue(XMLObject xmlObject) {
+		if (xmlObject == null) {
+			return null;
+		}
+		if (xmlObject instanceof XSAnyImpl) {
+			return ((XSAnyImpl) xmlObject).getTextContent();
```

Comment:
Since `XSAny` can have much more than text content, it makes me wonder whether we'll be losing information by only returning the text content.

I think let's skip this type for now (don't add it into the `Map`), we can add it later when we've got a better understanding of what use cases are still missing.

Reply:
Well... That's the only type of attribute values that we get from our server :)
We only get the name of an attribute value for some reason, without the type argument.

Reply:
I don't know how common this behavior is but [looks valid](https://lists.oasis-open.org/archives/saml-dev/201106/msg00001.html).

While we have an `XSAnyImpl` instance, it's QName is `{urn:oasis:names:tc:SAML:2.0:assertion}AttributeValue` and there is no marshaler for that, so I'm getting NPE both with the unit and e2e test.

This method was used in the [spring-security-saml](https://github.com/spring-projects/spring-security-saml/blob/708ab9839701afeafe3e430f4cb287cd58194c62/core/src/main/java/org/springframework/security/saml/SAMLCredential.java#L216) to get attribute value as a string.

Reply:
Perhaps something like this would be more robust:

```java
Map<QName, Marshaller> marshallers =
		XMLObjectProviderRegistrySupport.getMarshallerFactory().getMarshallers();
if (marshallers.containsKey(xmlObject.getSchemaType())) {
	return this.saml.serialize(xmlObject);
}
if (marshallers.containsKey(xmlObject.getElementQName())) {
	return this.saml.serialize(xmlObject);
}
return ((XSAnyImpl) xmlObject).getTextContent();
```
That way, an application can override the marshalling behavior if they need to by configuring OpenSAML's `MarshallerFactory`.

Reply:
I'm ok with that as long as I can unregister the marshaller if it gets registered by default in some future versions since text content satisfies my needs.